src/integrationTest/java/org/opensearch/security/http/AnonymousAuthenticationTest.java

```diff
@@ -80,7 +86,13 @@ public class AnonymousAuthenticationTest {
     public void shouldAuthenticate_positive_anonymousUser() {
         try (TestRestClient client = cluster.getRestClient()) {
 
-            TestRestClient.HttpResponse response = client.getAuthInfo();
+            Header anonyAuthHeader = new BasicHeader(
+                "Authorization",
+                "Basic "
+                    + Base64.getEncoder()
+                        .encodeToString((User.ANONYMOUS.getName() + ":" + randomAsciiAlphanumOfLength(8)).getBytes(StandardCharsets.UTF_8))
```

Comment:
anonymous auth currently creates an anonymous user in the backend to be added to thread context.

The root problem of this issue is that the current approach creates an anonymous user if no auth credentials were found, which means that the flow would never reach SAML IdP server since [this block](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/auth/BackendRegistry.java#L389-L400) will return `true` authentication is re-requested hence leading to `Invalid SAML Configuration` error since it was never able to reach SAML IdP server.

To solve this problem, we need to identify whether the request is coming as anonymous user or not. Following options were considered:
1. Add a header `{ _auth_request_type_: anonymous }` to incoming requests from anonymous. But this somehow got overwritten when a request was sent to read dashboards config. So this option was tabled.
2. Add auth creds from front-end to let backend know that this request is coming as anonymous user. This is the current approach and it allows differentiating auth provider requests with anonymous auth requests.

Reply:
Can we not just swap the order of processing by moving the block you linked to after any SAML auth attempts?

Reply:
already tried that. and that is the conversation Craig was referring to, to swap some code around to see if it works. My question is why are we opposed to using dummy creds coming from front-end, when we any way create an anonymous user on the backend? These creds are not requested from an user, instead they are generated on the fly when a user tries to log in as anonymous. I'm not using the password part of the creds, the only thing i need is the username part to match the anonymous user's name.

Reply:
Also, anonymous auth is considered a sub-part of basic auth (username + password) in the design, so I followed that same pattern in addressing this.

This would only cause some issues for API callers without passing creds. They would now have to pass in a basic auth header with [anonymous user name](https://github.com/opensearch-project/security/blob/d526c9f6c2a438c14db8b413148204510b9fe2e2/src/main/java/org/opensearch/security/user/User.java#L53) and a random string as password.

One possible solution, which is much larger than this bug fix is to re-write the authenticate method to ensure that anonymous auth is always checked at the end, but I'm not sure how would `Log in as anonymous` react in the case where multiple auth domains are enabled. How would we identify the type of login request?

Reply:
+1 on the API calls - I have no concerns from front end, since we are hooking up all the buttons anyway and can make the change appropriately, but not sure if it is considered a breaking change to require a username after this change